test/common/namespace_restoration.go

```diff
@@ -0,0 +1,282 @@
+	"testing"
+	"time"
+)
+
+const (
+	stageRestorationTimeOut = 30 * time.Minute
+	stageRetryInterval      = 15 * time.Second
+)
+
+type StageDeletion struct {
+	productStageName integreatlyv1alpha1.StageName
+	namespaces       []string
+	removeFinalizers func(ctx *TestingContext) error
+}
+
+var (
+	productStages = []StageDeletion{
+		{
+			productStageName: integreatlyv1alpha1.AuthenticationStage,
+			namespaces: []string{
+				RHSSOProductNamespace,
+				RHSSOOperatorNamespace,
+			},
+			removeFinalizers: func(ctx *TestingContext) error {
+				return removeKeyCloakFinalizers(ctx, RHSSOProductNamespace)
+			},
+		},
+		{
+			productStageName: integreatlyv1alpha1.CloudResourcesStage,
+			namespaces: []string{
+				CloudResourceOperatorNamespace,
+			},
+			removeFinalizers: func(ctx *TestingContext) error {
+				return nil
+			},
+		},
+		{
+			productStageName: integreatlyv1alpha1.MonitoringStage,
+			namespaces: []string{
+				MonitoringOperatorNamespace,
+				MonitoringFederateNamespace,
+			},
+			removeFinalizers: func(ctx *TestingContext) error {
+				return removeMonitoringFinalizers(ctx, MonitoringOperatorNamespace)
+			},
+		},
+		{
+			productStageName: integreatlyv1alpha1.ProductsStage,
+			namespaces: []string{
+				AMQOnlineOperatorNamespace,
+				ApicuritoProductNamespace,
+				ApicuritoOperatorNamespace,
+				CodeReadyProductNamespace,
+				CodeReadyOperatorNamespace,
+				FuseProductNamespace,
+				FuseOperatorNamespace,
+				RHSSOUserProductOperatorNamespace,
+				RHSSOUserOperatorNamespace,
+				// Removing 3scale product ns from test as "bundle exec rake zync:resync:domains" needs to be executed in system-app pod terminal to restore properly
+				// ThreeScaleProductNamespace,
+				// ThreeScaleOperatorNamespace,
+				UPSProductNamespace,
+				UPSOperatorNamespace,
+			},
+			removeFinalizers: func(ctx *TestingContext) error {
+				return removeKeyCloakFinalizers(ctx, RHSSOUserProductOperatorNamespace)
+			},
+		},
+		{
+			productStageName: integreatlyv1alpha1.SolutionExplorerStage,
+			namespaces: []string{
+				SolutionExplorerProductNamespace,
+				SolutionExplorerOperatorNamespace,
+			},
+			removeFinalizers: func(ctx *TestingContext) error {
+				return nil
+			},
+		},
+	}
+)
+
+func TestNamespaceRestoration(t *testing.T, ctx *TestingContext) {
+	for _, stage := range productStages {
+
+		// Delete all the namespaces defined in product stage
+		for _, nameSpace := range stage.namespaces {
+			nameSpaceForDeletion := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      nameSpace,
+					Namespace: nameSpace,
+				},
+			}
+
+			err := ctx.Client.Delete(goctx.TODO(), nameSpaceForDeletion)
+
+			if err != nil {
+				t.Fatalf("Error deleting the namespace - %s: %s", nameSpace, err)
+			}
+
+			t.Logf("Deleted %s namesapce", nameSpace)
+		}
+
+		// Sleep for 10 seconds before deleting finalizers as they can be re-added if namespace is not in a fully terminating state
+		time.Sleep(10 * time.Second)
```

Comment:
I can already imagine this may become a problem in the future, is there an easier way to pull the namespace status instead of relying on a fix time sleep?

https://github.com/integr8ly/integreatly-operator/blob/master/test/BEST_PRACTICES.md#dont-sleep

Reply:
Yeah, good point, I'll try some things either trying to get the namespace status or do a poll for deleting the finalizers :thinking: